### PR TITLE
Fix logging plugin crash unrecoverable

### LINKED ISSUE #36929: Broken Pipe errors are seen when an in use log driver plugin get killed
**Description**

Ideally, when a logdriver plugin get killed for some reason, oom for example, the stdout/stderr stream of a container which uses the plugin shall be blocked until the logdriver plugin gets recovered, or just fail fast. However, this is not the behavior of the current logdriver plugin implementation, when a logdriver plugin crashed, if any container is streaming logs to this plugin,  the container will keep running without blocking, 'Broken pipe' errors can be seen from docker engine logs which indicates there's unexpected data loss.  

**Steps to reproduce the issue:**
1. Set up a docker logdriver plugin to use, with the following script
```
#!/bin/sh
git clone git@github.com:cpuguy83/docker-log-driver-test.git
cd docker-log-driver-test
# fix the broken dependency by removing two lines of code
sed -i '/msg.Partial/d' driver.go
docker build -f Dockerfile.build . -t test-log-driver
docker run --name tmp test-log-driver:latest
mkdir -p plugin/rootfs
cp config.json plugin
docker export tmp | tar x -C plugin/rootfs
docker plugin create test-driver ./plugin
docker plugin set test-driver:latest LOG_LEVEL=debug
docker plugin enable test-driver:latest
```
2. Start a docker container with this logdriver
```
docker run  --name test-log-driver --rm --log-driver test-driver:latest alpine /bin/sh -c 'for i in `seq 1 10000`; do echo "log item $i"; sleep 1; done'
```
3. On a different terminal, identify the process id of the logdriver application on the host
```
> ps aux  | grep docker-log-driver
# some thing like below can be seen
root     25588  0.0  0.0 283988 11144 ?        Ssl  Apr13   0:01 /usr/bin/docker-log-driver
```
then kill this process with SIGKILL. 

**Describe the results you received:**
Open journalctl to check logs
```
 journalctl -f
```
error message with Broken Pipe like the following can be seen
```
Apr 13 04:23:27 hostname dockerd[1247]: time="2018-04-13T04:23:27.602363175-07:00" level=error msg="Failed to log msg \"\" for logger test-driver:latest: write /proc/self/fd/40: broken pipe"
Apr 13 04:23:27 hostname dockerd[1247]: time="2018-04-13T04:23:27.604785825-07:00" level=error msg="Failed to log msg \"\" for logger test-driver:latest: write /proc/self/fd/40: broken pipe"
Apr 13 04:23:27 hostname dockerd[1247]: time="2018-04-13T04:23:27.607369644-07:00" level=error msg="Failed to log msg \"\" for logger test-driver:latest: write /proc/self/fd/40: broken pipe"
```

And the container is still running

**Describe the results you expected:**
Shall not see broken pipe errors from docker-engine side.  Since this is a toy log driver plugin, it did not have any recovery functionality. In this case container logs shall be just blocked which behaviors like downstream is down.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Docker version 17.05.0-ce, build 89658be
```

**Output of `docker info`:**

```
Containers: 8
 Running: 2
 Paused: 0
 Stopped: 6
Images: 53
Server Version: 17.05.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 233
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9048e5e50717ea4497b757314bad98ea3763c145
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-28-generic
Operating System: Ubuntu 16.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.59GiB
Name: poincare
ID: BAWE:EBZ7:W5PE:KYAZ:7APE:VEOC:QOYO:GS37:D75R:NVTD:2D22:KCTE
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: fivesheep
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Potential solution**

When open a fifo for write from [plugin.go](https://github.com/moby/moby/blob/53683bd8326b988977650337ee43b281d2830076/daemon/logger/plugin.go#L72),  it shall also open the same fifo in read mode at the same time. I was able to simulate this behavior by using an external python script to open the pipe in read mode but not reading any data, and this additional dummy reader won't cause any issue to the actual reader. The same solution can be seen in [containerd](https://github.com/containerd/containerd/blob/209a7fc3e4a32ef71a8c7b50c68fc8398415badf/linux/proc/io.go#L86-L110)

### PR DESCRIPTION
**Pull request by cpuguy83**
In cases where a logging plugin has crashed when the daemon tries to
copy the container stdio to the logging plugin it returns a broken pipe
error and any log entries that occur while the plugin is down are lost.

Fix this by opening read+write in the daemon so logs are not lost while
the plugin is down.

Fixes #36929

### CODE REVIEW

**Comment by thaJeztah** [comment]
hmm 

```
daemon/logger/fluentd/fluentd.go:166:5:warning: invalid operation: msg (variable of type *github.com/docker/docker/daemon/logger.Message) has no field or method Partial (interfacer)
21:34:27 daemon/logger/fluentd/fluentd.go:166:5:warning: invalid operation: msg (variable of type *github.com/docker/docker/daemon/logger.Message) has no field or method Partial (unconvert)
21:34:27 daemon/logger/fluentd/fluentd.go:166:5:warning: invalid operation: msg (variable of type *github.com/docker/docker/daemon/logger.Message) has no field or method Partial (gosimple)
```

**Comment by thaJeztah** [comment]
dang; I guess i broke the build with https://github.com/moby/moby/pull/36159

**Comment by thaJeztah** [comment]
fixing: https://github.com/moby/moby/pull/37029

**Comment by cpuguy83** [comment]
This is ready.

**Review by vdemeester — APPROVED** [review]
LGTM :tiger:

File: daemon/logger/plugin_unix.go [hunk 1]
```diff
@@ -12,7 +12,10 @@ import (
 )
 
 func openPluginStream(a *pluginAdapter) (io.WriteCloser, error) {
-	f, err := fifo.OpenFifo(context.Background(), a.fifoPath, unix.O_WRONLY|unix.O_CREAT|unix.O_NONBLOCK, 0700)
+	// open with read/write so avoid broken pipes in case of a plugin failure.
```
Comment (anusha-ragunathan) [inline daemon/logger/plugin_unix.go]: typo: "so" -> "to".
Also, open with write is obvious. the read is the non-obvious part and causing the broken pipe. So maybe reword to emphasize that?

File: daemon/logger/plugin_unix.go [hunk 2]
```diff
@@ -12,7 +12,10 @@ import (
...
+	// It is up to the plugin to keep track of pipes that it should re-attach to, however.
+	// If the plugin doesn't open for reads, then the container will block once the pipe is full.
+	f, err := fifo.OpenFifo(context.Background(), a.fifoPath, unix.O_RDWR|unix.O_CREAT|unix.O_NONBLOCK, 0700)
```
Comment (anusha-ragunathan) [inline daemon/logger/plugin_unix.go]: Not strictly related to this PR, but containerd repos all use OpenFifo with permMode set to `0` and not `0700`. Any ideas why?

File: integration/plugin/logging/cmd/close_on_start/main.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintln(w, `{}`)
+	})
+	server := http.Server{
+		Addr:    l.Addr().String(),
+		Handler: mux,
+	}
+
+	server.Serve(l)
```
Comment (anusha-ragunathan) [inline integration/plugin/logging/cmd/close_on_start/main.go]: missing error check. Also run in a goroutine?

Comment (cpuguy83) [reply integration/plugin/logging/cmd/close_on_start/main.go]: This will always return an error and will exit either way with the plugin manager trying to restart it.

File: daemon/logger/plugin_unix.go (hunk 2 above)
Comment (cpuguy83) [reply daemon/logger/plugin_unix.go]: Not sure, some of it is `0` and some is `0700`

```
vendor/github.com/containerd/containerd/cio/io_unix.go:114:      if f.Stdin, err = fifo.OpenFifo(ctx, fifos.Stdin, syscall.O_WRONLY|syscall.O_CREAT|syscall.O_NONBLOCK, 0700); err != nil {
vendor/github.com/containerd/containerd/cio/io_unix.go:124:      if f.Stdout, err = fifo.OpenFifo(ctx, fifos.Stdout, syscall.O_RDONLY|syscall.O_CREAT|syscall.O_NONBLOCK, 0700); err != nil {
vendor/github.com/containerd/containerd/cio/io_unix.go:134:      if f.Stderr, err = fifo.OpenFifo(ctx, fifos.Stderr, syscall.O_RDONLY|syscall.O_CREAT|syscall.O_NONBLOCK, 0700); err != nil {
vendor/github.com/containerd/containerd/linux/proc/exec.go:166:      sc, err := fifo.OpenFifo(ctx, e.stdio.Stdin, syscall.O_WRONLY|syscall.O_NONBLOCK, 0)
vendor/github.com/containerd/containerd/linux/proc/init.go:203:      sc, err := fifo.OpenFifo(context, r.Stdin, syscall.O_WRONLY|syscall.O_NONBLOCK, 0)
vendor/github.com/containerd/containerd/linux/proc/init_state.go:202:        sc, err := fifo.OpenFifo(ctx, sio.Stdin, syscall.O_WRONLY|syscall.O_NONBLOCK, 0)
vendor/github.com/containerd/containerd/linux/proc/io.go:69:     fw, err := fifo.OpenFifo(ctx, name, syscall.O_WRONLY, 0)
vendor/github.com/containerd/containerd/linux/proc/io.go:73:     fr, err := fifo.OpenFifo(ctx, name, syscall.O_RDONLY, 0)
vendor/github.com/containerd/containerd/linux/proc/io.go:83: f, err := fifo.OpenFifo(ctx, stdin, syscall.O_RDONLY, 0)
vendor/github.com/containerd/containerd/linux/shim/service_linux.go:45:      in, err := fifo.OpenFifo(ctx, stdin, syscall.O_RDONLY, 0)
vendor/github.com/containerd/containerd/linux/shim/service_linux.go:58:  outw, err := fifo.OpenFifo(ctx, stdout, syscall.O_WRONLY, 0)
vendor/github.com/containerd/containerd/linux/shim/service_linux.go:62:  outr, err := fifo.OpenFifo(ctx, stdout, syscall.O_RDONLY, 0)
vendor/github.com/containerd/containerd/linux/shim/service_unix.go:36:       in, err := fifo.OpenFifo(ctx, stdin, syscall.O_RDONLY, 0)
vendor/github.com/containerd/containerd/linux/shim/service_unix.go:49:   outw, err := fifo.OpenFifo(ctx, stdout, syscall.O_WRONLY, 0)
vendor/github.com/containerd/containerd/linux/shim/service_unix.go:53:   outr, err := fifo.OpenFifo(ctx, stdout, syscall.O_RDONLY, 0)
```

File: daemon/logger/plugin_unix.go (hunk 1 above)
Comment (cpuguy83) [reply daemon/logger/plugin_unix.go]: Will do.

File: integration/plugin/logging/cmd/close_on_start/main.go (hunk 3 above)
Comment (anusha-ragunathan) [reply integration/plugin/logging/cmd/close_on_start/main.go]: Better to add a comment in L23 explaining that not suppling a file there is intentional to cause the plugin to error out.

**Comment by anusha-ragunathan** [comment]
LGTM

**Review by thaJeztah — APPROVED** [review]
LGTM

ignoring codecov; no idea why it's showing 0%
